src/network/operator.rs

```diff
@@ -0,0 +1,335 @@
+                } else {
+                    warn!("[PoolRegister] No current block template exists");
+                }
+            }
+            OperatorRequest::PoolResponse(peer_ip, block, prover_address) => {
+                if let Some(block_template) = &*self.block_template.read().await {
+                    // Check that the block is relevant.
+                    if self.ledger_reader.latest_block_height().saturating_add(1) != block.height() {
+                        warn!("[PoolResponse] Peer {} sent a stale candidate block.", peer_ip);
+                        return;
+                    }
+
+                    // Retrieve the coinbase transaction records.
+                    let coinbase_records = match block.to_coinbase_transaction() {
+                        Ok(transaction) => {
+                            // Ensure the owner of the coinbase transaction in the block is the operator address.
+                            let coinbase_records: Vec<Record<N>> =
+                                transaction.to_records().filter(|r| Some(r.owner()) == self.address).collect();
+                            if coinbase_records.len() == 0 {
+                                warn!("[ProposedBlock] Peer {} sent a candidate block with an incorrect owner.", peer_ip);
+                                return;
+                            }
+                            coinbase_records
+                        }
+                        Err(error) => {
+                            warn!("[PoolResponse] {}", error);
+                            return;
+                        }
+                    };
+
+                    // Ensure the block contains a difficulty that is at least the share difficulty.
+                    let proof_difficulty = {
+                        let proof_bytes = match block.header().proof() {
+                            Some(proof) => match proof.to_bytes_le() {
+                                Ok(bytes) => bytes,
+                                Err(error) => {
+                                    warn!("[PoolResponse] {}", error);
+                                    return;
+                                }
+                            },
+                            None => {
+                                warn!("[PoolResponse] Peer {} sent a candidate block with a missing proof.", peer_ip);
+                                return;
+                            }
+                        };
+
+                        let proof_difficulty = sha256d_to_u64(&proof_bytes);
+                        let share_difficulty = {
+                            let mut provers = self.provers.write().await;
+                            match provers.get(&prover_address) {
+                                Some((_, share_difficulty, _)) => *share_difficulty,
+                                None => {
+                                    provers.insert(prover_address, (Instant::now(), BASE_SHARE_DIFFICULTY, 0));
```

Comment:
general note: if we don't have to be precise to the nanosecond with the timestamp, I'd suggest calling `Instant::now()` before a lock is acquired in order to limit its scope